### PR TITLE
resolved: properly handle stub replies for chains of multiple CNAMEs

### PR DESCRIPTION
**Pull request by poettering**
A very thorough fix for #18819

### CODE REVIEW

File: src/resolve/resolved-dns-stub.c [hunk 1]
```diff
@@ -162,79 +162,88 @@ static int dns_stub_collect_answer_by_question(
...
         _cleanup_(dns_resource_key_unrefp) DnsResourceKey *redirected_key = NULL;
+        unsigned n_cname_redirects = 0;
         DnsAnswerItem *item;
         int r;
 
         assert(reply);
 
-        /* Copies all RRs from 'answer' into 'reply', if they match 'question'. */
+        /* Copies all RRs from 'answer' into 'reply', if they match 'question'. There might be direct and
+         * indirect matches (i.e. via CNAME/DNAME). If the have an indirect one, remember where we need to
```
Comment (keszybz) [inline src/resolve/resolved-dns-stub.c]: "the have" ?

File: src/resolve/resolved-dns-stub.c [hunk 2]
```diff
@@ -162,79 +162,88 @@ static int dns_stub_collect_answer_by_question(
...
                                 if (r == -EUNATCH)
                                         continue; /* Not a CNAME/DNAME or doesn't match */
                                 if (r < 0)
                                         return r;
 
-                                dns_resource_key_unref(redirected_key);
+                                /* Oh, wow, this is a redirect. Let's remember where this points, and store
+                                 * it in 'next_redirected_key'. Once we finished iterating through the rest
+                                 * of the RR's we'll start again, with the redirected RR. */
+
```
Comment (keszybz) [inline src/resolve/resolved-dns-stub.c:215]: Yep, this sounds like it should fix the issue.

File: src/resolve/resolved-dns-answer.c [hunk 3]
```diff
@@ -640,6 +640,31 @@ int dns_answer_remove_by_rr(DnsAnswer **a, DnsResourceRecord *rm) {
...
+        DNS_ANSWER_FOREACH_ITEM(item, b) {
+
+                if (prev && dns_resource_key_equal(item->rr->key, prev)) /* Skip this one, we already looked at it */
+                        continue;
+
+                r = dns_answer_remove_by_key(a, item->rr->key);
+                if (r < 0)
+                        return r;
+
+                /* Let's remember this entrie's RR key, to optimize the loop a bit: if we have an RRset with
```
Comment (keszybz) [inline src/resolve/resolved-dns-answer.c]: entry's

File: src/resolve/resolved-dns-answer.c [hunk 4]
```diff
@@ -640,6 +640,31 @@ int dns_answer_remove_by_rr(DnsAnswer **a, DnsResourceRecord *rm) {
...
+                 * more than one item then we don#t need to remove the key multiple times */
```
Comment (keszybz) [inline src/resolve/resolved-dns-answer.c]: don't

File: src/resolve/resolved-dns-stub.c [hunk 5]
```diff
@@ -574,6 +574,22 @@ static int dns_stub_reply_with_edns0_do(DnsQuery *q) {
                   DNS_PACKET_CD(q->request_packet));     /* … or client set CD */
 }
 
+static void dns_stub_suppress_duplicate_section_rrs(DnsQuery *q) {
+        /* If we follow a CNAME/DNAME chain we might end up populating our sections with redundant RRs
+         * because we built up the sections from multiple reply packets (one from each CNAME/DNAME chain
+         * element). e.g. it could be that an RR that was included in the first reply's additional section
```
Comment (keszybz) [inline src/resolve/resolved-dns-stub.c]: "E.g."

File: src/resolve/resolved-dns-stub.c [hunk 6]
```diff
@@ -574,6 +574,22 @@ static int dns_stub_reply_with_edns0_do(DnsQuery *q) {
...
+         * ends up being relevant as main answer in a subsequent reply in the chain. Let's clean this up, and
+         * remove everything from the "higher priority" sections from the "lower priority" sections if they
```
Comment (keszybz) [inline src/resolve/resolved-dns-stub.c]: "in the "higher priority" sections from the "lower priority" sections"
~~if they exist in both~~

**Review by keszybz — COMMENTED** [review]
Looks great. Some typos in comments…

I didn't test this yet, but it looks like it should fix the issues we have seen.

**Comment by poettering** [comment]
Force pushed a new version with the suggested typo fixes. Upgrading green label.

**Pull request merged**